### ISSUE TITLE
HttpClientRequest not mock-/stubable

### ISSUE THREAD

**Issue by leslie-leder** [ReactiveX/RxNetty#529]
In order to test my specific http tasks I would like to know whether they configure the HttpClientRequest properly, e.g. which headers are set, which query, uri, etc.
There for it would be quite handy if I could stub or mock the HttpClientRequest instance which is passed to my specific http task under test. 

```
val requestStub = stub[HttpClientRequest[ByteBuf, ByteBuf]]
val response = MyTask.apply(requestStub)`
```

This is prevented by the HttpClientRequest's inheritance from Observable which has no default constructor but has final methods.

This urges me to always test with a real http environment. Which is not to hard when using 
[WireMock](http://wiremock.org/). But still its is some overhead. 
For some scenarios (e.g. http server notifications) I do not have an idea how to test it without mock/stub. 

Do you have a proposal how to test client code without setting up real http servers (embedded or not)?

Possibly the inheritance on Obsevable could be questioned? For all request sending data down to the server the Observable nature of HttpClientRequest it is anyway of no use. They get a response observable from the  _writeContent(rx.Observable<I> contentSource)_ method.

Working with Observables in tests is in general quite cool but the inheritance seems a bit problematic to me.

**Comment by leslie-leder** [comment]
More precisely I would like to test whether the following Http request code converts correctly from ByteBuf content to String content and whether it propagates Http errors properly over onError invocation.

```
class HttpEventStreamConsumer(client: HttpClient[ByteBuf, ByteBuf]) {
  def apply(uri: String): Observable[String] =
    client.createGet(uri).asScala
      .failOnHttpError()
      .concatMap { response =>
        response.getContent().asScala.map { byteBuf =>
          val content = byteBuf.toString(StandardCharsets.UTF_8)
          if (byteBuf.refCnt() > 0)
            byteBuf.release()
          content
        }
      }
}
```

Intuitively I would create a client mock which returns a client request, ...

```
class HttpEventStreamConsumerTest extends FreeSpec with MockFactory {
  "convert ByteBuf content to String content" in {
    val clientStub = stub[HttpClient[ByteBuf, ByteBuf]]

    val requestStub = stub[HttpClientRequest[ByteBuf, ByteBuf]]

    (clientStub.createGet _) when ("/test-uri") returns requestStub

    new HttpEventStreamConsumer(clientStub)
      .apply("/test-uri")
      .subscribe()
    ...
  }
}
```

Scala Mock cannot stub this requestStub of type HttpClientRequest and I guess Mockito would also complain. 

How can I inject an HttpClientRequest into my object under test, so I can see that my class works correct on a sequence of ByteBuf's?

For now I use WireMock in order to test on Http level. But I cannot persuade it to send multiple text chunks in fashion that I receive multiple ByteBuf content items from the HttpResponse.getContent stream.

**Comment by NiteshKant** [comment]
@leslie-leder yes having `HttpClientRequest` extend `Observable` has this annoying side-effect of difficulty in testing. OTOH, it is quiet nice in terms of usability as you can do things like this:

``` java
HttpClient.newClient(serverAddress)
                .createGet("/hello")
                .subscribe(...)
```

which is pretty terse and cool 😄 instead of introducing another entity of `HttpClientRequest` which is necessarily created out-of-band and passed to the client to create an `Observable` for response.

I tend to believe that having an easier API to interact at runtime is a good tradeoff at the cost of making tests a bit more complex.

I was intending to provide a `TestHttpClient`/`TestTcpClient` that stubs out everything with the same `Client` API but provides input and output as a queues to which someone can write and read from! Somewhat similar to netty's `EmbeddedChannel`. If you want to take a stab at it and send a PR, you are welcome 😄
